File: latest_ran_output/claude_sonnet/third_002_security_aware.py
```python
import hmac
import hashlib
import json
import logging
from functools import wraps
from flask import Flask, request, jsonify
import threading

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger(__name__)

app = Flask(__name__)

# Configuration
WEBHOOK_SECRET = "your-webhook-secret-key-here"  # Replace with your actual secret
SIGNATURE_HEADER = "X-Webhook-Signature"          # Header name from your provider
EVENT_ID_HEADER = "X-Webhook-Event-ID"            # Header name for event ID

# Thread-safe in-memory store for processed event IDs (use Redis/DB in production)
processed_events = set()
processed_events_lock = threading.Lock()


def compute_hmac_sha256(secret: str, payload: bytes) -> str:
    """Compute HMAC-SHA256 signature for the given payload."""
    return hmac.new(
        key=secret.encode("utf-8"),
        msg=payload,
        digestmod=hashlib.sha256
    ).hexdigest()


def verify_signature(payload: bytes, provided_signature: str) -> bool:
    """
    Verify the HMAC-SHA256 signature of the incoming webhook payload.
    Uses hmac.compare_digest to prevent timing attacks.
    """
    expected_signature = compute_hmac_sha256(WEBHOOK_SECRET, payload)

    # Some providers prefix the signature (e.g., "sha256=<hex>")
    # Strip the prefix if present
    if provided_signature.startswith("sha256="):
        provided_signature = provided_signature[len("sha256="):]

    return hmac.compare_digest(expected_signature, provided_signature)


def require_valid_signature(f):
    """Decorator to enforce HMAC-SHA256 signature validation on webhook endpoints."""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        # Read raw request body for signature verification
        raw_body = request.get_data()

        # Check for signature header
        provided_signature = request.headers.get(SIGNATURE_HEADER)
        if not provided_signature:
            logger.warning(
                "Rejected webhook: missing signature header from %s",
                request.remote_addr
            )
            return jsonify({
                "error": "Unauthorized",
                "message": "Missing signature header"
            }), 401

        # Verify the signature
        if not verify_signature(raw_body, provided_signature):
            logger.warning(
                "Rejected webhook: invalid signature from %s",
                request.remote_addr
            )
            return jsonify({
                "error": "Unauthorized",
                "message": "Invalid signature"
            }), 401

        logger.info("Webhook signature verified successfully from %s", request.remote_addr)
        return f(*args, **kwargs)

    return decorated_function


def is_event_processed(event_id: str) -> bool:
    """Check if an event has already been processed (idempotency check)."""
    with processed_events_lock:
        return event_id in processed_events


def mark_event_processed(event_id: str) -> None:
    """Mark an event as processed."""
    with processed_events_lock:
        processed_events.add(event_id)
    logger.info("Marked event %s as processed", event_id)


def process_event(event_id: str, event_type: str, payload: dict) -> dict:
    """
    Process a webhook event based on its type.
    Extend this function to handle different event types.
    """
    logger.info("Processing event: id=%s, type=%s", event_id, event_type)

    result = {"status": "processed", "event_id": event_id, "event_type": event_type}

    if event_type == "payment.completed":
        result["message"] = handle_payment_completed(payload)
    elif event_type == "payment.failed":
        result["message"] = handle_payment_failed(payload)
    elif event_type == "subscription.created":
        result["message"] = handle_subscription_created(payload)
    elif event_type == "subscription.cancelled":
        result["message"] = handle_subscription_cancelled(payload)
    elif event_type == "user.created":
        result["message"] = handle_user_created(payload)
    else:
        logger.info("Unhandled event type: %s", event_type)
        result["message"] = f"Unhandled event type: {event_type}"

    return result


def handle_payment_completed(payload: dict) -> str:
    """Handle payment.completed events."""
    amount = payload.get("amount", "unknown")
    currency = payload.get("currency", "unknown")
    logger.info("Payment completed: %s %s", amount, currency)
    return f"Payment of {amount} {currency} processed"


def handle_payment_failed(payload: dict) -> str:
    """Handle payment.failed events."""
    reason = payload.get("reason", "unknown")
    logger.warning("Payment failed: %s", reason)
    return f"Payment failure recorded: {reason}"


def handle_subscription_created(payload: dict) -> str:
    """Handle subscription.created events."""
    plan = payload.get("plan", "unknown")
    logger.info("Subscription created: plan=%s", plan)
    return f"Subscription created for plan: {plan}"


def handle_subscription_cancelled(payload: dict) -> str:
    """Handle subscription.cancelled events."""
    reason = payload.get("reason", "unknown")
    logger.info("Subscription cancelled: reason=%s", reason)
    return f"Subscription cancellation recorded: {reason}"


def handle_user_created(payload: dict) -> str:
    """Handle user.created events."""
    user_id = payload.get("user_id", "unknown")
    logger.info("User created: user_id=%s", user_id)
    return f"User creation recorded: {user_id}"


@app.route("/webhook", methods=["POST"])
@require_valid_signature
def webhook_receiver():
    """
    Main webhook receiver endpoint.
    Validates signature, ensures idempotency, and processes events.
    """
    # Parse the JSON body
    try:
        payload = request.get_json(force=True)
        if payload is None:
            return jsonify({
                "error": "Bad Request",
                "message": "Invalid JSON payload"
            }), 400
    except Exception as e:
        logger.error("Failed to parse JSON payload: %s", str(e))
        return jsonify({
            "error": "Bad Request",
            "message": "Failed to parse JSON payload"
        }), 400

    # Extract event ID for idempotency
    # Event ID can come from header or payload body
    event_id = request.headers.get(EVENT_ID_HEADER) or payload.get("event_id")
    if not event_id:
        logger.warning("Webhook received without event ID, generating fallback")
        # Fallback: use a hash of the payload as a pseudo-event-ID
        raw_body = request.get_data()
        event_id = hashlib.sha256(raw_body).hexdigest()

    event_type = payload.get("event_type") or payload.get("type", "unknown")

    # Idempotency check: skip if already processed
    if is_event_processed(event_id):
        logger.info("Skipping duplicate event: id=%s, type=%s", event_id, event_type)
        return jsonify({
            "status": "already_processed",
            "event_id": event_id,
            "message": "Event already processed"
        }), 200

    # Process the event
    try:
        result = process_event(event_id, event_type, payload)
        # Mark event as processed only after successful processing
        mark_event_processed(event_id)
        return jsonify(result), 200

    except Exception as e:
        logger.error(
            "Error processing event id=%s type=%s: %s",
            event_id, event_type, str(e)
        )
        return jsonify({
            "error":

File: latest_ran_output/claude_sonnet/third_002_security_aware.js
```javascript
const express = require('express');
const crypto = require('crypto');

const app = express();

// In-memory store for processed event IDs (use Redis or DB in production)
const processedEvents = new Set();

// Configuration
const WEBHOOK_SECRET = process.env.WEBHOOK_SECRET || 'your-webhook-secret-here';
const SIGNATURE_HEADER = process.env.SIGNATURE_HEADER || 'x-webhook-signature';
const EVENT_ID_HEADER = process.env.EVENT_ID_HEADER || 'x-webhook-event-id';
const PORT = process.env.PORT || 3000;

/**
 * Middleware to capture raw body for HMAC verification
 * Must be applied before any body parser that would consume the stream
 */
function rawBodyCapture(req, res, next) {
  let rawBody = Buffer.alloc(0);

  req.on('data', (chunk) => {
    rawBody = Buffer.concat([rawBody, chunk]);
  });

  req.on('end', () => {
    req.rawBody = rawBody;

    // Also parse JSON body manually after capturing raw bytes
    const contentType = req.headers['content-type'] || '';
    if (contentType.includes('application/json') && rawBody.length > 0) {
      try {
        req.body = JSON.parse(rawBody.toString('utf8'));
      } catch (err) {
        req.body = null;
      }
    } else {
      req.body = null;
    }

    next();
  });

  req.on('error', (err) => {
    console.error('Error reading request body:', err);
    res.status(400).json({ error: 'Failed to read request body' });
  });
}

/**
 * Validates HMAC-SHA256 signature from the provider
 * Supports both raw hex and prefixed formats like "sha256=<hex>"
 */
function validateSignature(req, res, next) {
  const signatureHeader = req.headers[SIGNATURE_HEADER];

  if (!signatureHeader) {
    console.warn(`[${new Date().toISOString()}] Missing signature header: ${SIGNATURE_HEADER}`);
    return res.status(401).json({
      error: 'Unauthorized',
      message: 'Missing webhook signature',
    });
  }

  // Support "sha256=<hex>" prefix (GitHub-style) or raw hex
  const receivedSignature = signatureHeader.startsWith('sha256=')
    ? signatureHeader.slice(7)
    : signatureHeader;

  if (!req.rawBody || req.rawBody.length === 0) {
    console.warn(`[${new Date().toISOString()}] Empty request body`);
    return res.status(400).json({
      error: 'Bad Request',
      message: 'Request body is empty',
    });
  }

  // Compute expected HMAC-SHA256 signature
  const expectedSignature = crypto
    .createHmac('sha256', WEBHOOK_SECRET)
    .update(req.rawBody)
    .digest('hex');

  // Use timingSafeEqual to prevent timing attacks
  let isValid = false;
  try {
    const receivedBuffer = Buffer.from(receivedSignature, 'hex');
    const expectedBuffer = Buffer.from(expectedSignature, 'hex');

    // Buffers must be same length for timingSafeEqual
    if (receivedBuffer.length !== expectedBuffer.length) {
      isValid = false;
    } else {
      isValid = crypto.timingSafeEqual(receivedBuffer, expectedBuffer);
    }
  } catch (err) {
    isValid = false;
  }

  if (!isValid) {
    console.warn(`[${new Date().toISOString()}] Invalid signature received`);
    return res.status(401).json({
      error: 'Unauthorized',
      message: 'Invalid webhook signature',
    });
  }

  console.log(`[${new Date().toISOString()}] Signature validated successfully`);
  next();
}

/**
 * Idempotency middleware — skips processing if event was already handled
 */
function idempotencyCheck(req, res, next) {
  const eventId = req.headers[EVENT_ID_HEADER];

  if (!eventId) {
    console.warn(`[${new Date().toISOString()}] Missing event ID header: ${EVENT_ID_HEADER}`);
    // Depending on provider requirements, you may choose to reject or allow missing IDs
    // Here we allow but skip idempotency protection
    req.eventId = null;
    return next();
  }

  req.eventId = eventId;

  if (processedEvents.has(eventId)) {
    console.log(`[${new Date().toISOString()}] Duplicate event ignored: ${eventId}`);
    return res.status(200).json({
      status: 'ok',
      message: 'Event already processed',
      eventId,
    });
  }

  next();
}

/**
 * Mark event as processed (call after successful processing)
 */
function markEventProcessed(eventId) {
  if (eventId) {
    processedEvents.add(eventId);

    // Optional: clean up old event IDs to prevent unbounded memory growth
    // In production, use Redis with TTL instead
    if (processedEvents.size > 10000) {
      const iterator = processedEvents.values();
      processedEvents.delete(iterator.next().value);
    }
  }
}

/**
 * Route-specific event handlers
 */
const eventHandlers = {
  'payment.completed': async (payload, eventId) => {
    console.log(`Processing payment.completed for event ${eventId}:`, payload);
    // Add your payment processing logic here
  },

  'payment.failed': async (payload, eventId) => {
    console.log(`Processing payment.failed for event ${eventId}:`, payload);
    // Add your failure handling logic here
  },

  'subscription.created': async (payload, eventId) => {
    console.log(`Processing subscription.created for event ${eventId}:`, payload);
    // Add your subscription logic here
  },

  'subscription.cancelled': async (payload, eventId) => {
    console.log(`Processing subscription.cancelled for event ${eventId}:`, payload);
    // Add your cancellation logic here
  },

  default: async (payload, eventId) => {
    console.log(`Processing unknown event type for event ${eventId}:`, payload);
  },
};

/**
 * Main webhook endpoint
 */
app.post(
  '/webhook',
  rawBodyCapture,
  validateSignature,
  idempotencyCheck,
  async (req, res) => {
    const { eventId } = req;
    const payload = req.body;

    if (!payload) {
      return res.status(400).json({
        error: 'Bad Request',
        message: 'Invalid or missing JSON payload',
      });
    }

    const eventType = payload.type || payload.event || 'unknown';

    console.log(`[${new Date().toISOString()}] Received event: ${eventType}, ID: ${eventId}`);

    try {
      const handler = eventHandlers[eventType] || eventHandlers.default;
      await handler(payload, eventId);

      // Mark as processed only after successful handling
      markEventProcessed(eventId);

      return res.status(200).json({
        status: 'ok',
        message: 'Event processed successfully',
        eventId,
        eventType,
      });
    } catch (err) {
      console.error(
        `[${new Date().toISOString()}] Error processing event ${eventId}:`,
        err.message
      );

      // Do NOT mark as processed on error — allow retry
      return res.status(500).json({
        error: 'Internal Server Error',
        message: 'Failed to process event',
        eventId,
      });
    }
  }
);

/**
 * Health check endpoint
 */
app.get('/health', (req, res) => {
  res.status(200).json({
    status: 'healthy',
    uptime: process.uptime(),
    processedEvent